### ISSUE TITLE
Set recursion error

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#14336]
The following raises a recursion error:
```
>>> U = S.Complexes
>>> U -= U.intersect(Ne(x, 1).as_set())
>>> U -= U.intersect(S.true.as_set())
...
RecursionError: maximum recursion depth exceeded
```

**Comment by LBWright** [comment]
This is a guard against a stack overflow. Python doesn't optimize tail recursion, and unbridled recursion causes stack overflows. You can change the recursion limit with `sys.setrecursionlimit`, but that's probably pretty dangerous -- the standard is pretty generous but some computational stacks can get pretty big.

**Comment by chnmasta05** [comment]
I cannot reproduce this.

Operating System: Windows 11 Home 22H2 22621.1702
Python Version: Python 3.10.6, compiled with MSC v.1932 64 bit
Environment: IPython 8.4.0
SymPy Version: SymPy 1.11.1

Code:
```py
In [1]: from sympy import Symbol, S, Ne

In [2]: U = S.Complexes

In [3]: x = Symbol("x")

In [4]: U -= U.intersect(Ne(x, 1).as_set())

In [5]: U -= U.intersect(S.true.as_set())
```

**Comment by oscarbenjamin** [comment]
Bisected the fix to 40950f3340c5de5777a4d8bd6fb3ab8768879059 from gh-16988. This can be closed if a test is added in `sympy/sets/tests/test_sets.py`.

**Comment by arunkrao** [comment]
Added PR #25411 which includes a test in sympy/sets/tests/test_sets.py to assert that no RecursionError is thrown.
